### PR TITLE
feat: added pre-wired rule targetting HPCS in fscloud module

### PR DESCRIPTION
**Pull request by Ak-sky**
### Description

Added support and pre-wired rule for HPCS

fixes https://github.com/terraform-ibm-modules/terraform-ibm-cbr/issues/300

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

##### Release notes content

- With this release, pre-wired rules will be created using the HPCS (hs-crypto) as target service instead of Key Protect.

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- When merging, use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: examples/fscloud/variables.tf [hunk 1]
```diff
@@ -69,3 +69,15 @@ variable "allow_at_to_cos" {
   description = "Set rule for Activity Tracker to COS, default is true"
   default     = true
 }
+
+variable "kms" {
+  type        = list(string)
+  description = "List specific endpoint types for target services, valid values for endpoints are 'public', 'private' or 'direct'"
```
Comment (vburckhardt) [inline examples/fscloud/variables.tf:75]: Seems to be irrelevant here - copy pasted from somewhere else?

File: modules/fscloud/variables.tf [hunk 2]
```diff
@@ -195,3 +195,15 @@ variable "location" {
   description = "The region in which the network zone is scoped"
   default     = null
 }
+
+variable "kms" {
+  type        = list(string)
+  description = "List specific endpoint types for target services, valid values for endpoints are 'public', 'private' or 'direct'"
```
Comment (vburckhardt) [inline modules/fscloud/variables.tf:201]: Same as above

File: modules/fscloud/variables.tf [hunk 3]
```diff
@@ -195,3 +195,15 @@ variable "location" {
...
+  default     = ["kms", "hs-crypto"]
```
Comment (vburckhardt) [inline modules/fscloud/variables.tf:202]: I would default to hs-crypto only. Having both makes little sense in real world as well.

File: modules/fscloud/README.md [hunk 4]
```diff
@@ -1,13 +1,13 @@
 # Pre-wired CBR configuration for FS Cloud
 
 This module creates default coarse-grained CBR rules in a given account following a "secure by default" approach - that is: deny all flows by default, except known documented communication in the [Financial Services Cloud Reference Architecture](https://cloud.ibm.com/docs/framework-financial-services?topic=framework-financial-services-vpc-architecture-about):
-- COS -> KMS
-- Block storage -> KMS
-- ROKS -> KMS
+- COS -> KMS and HPCS
```
Comment (vburckhardt) [inline modules/fscloud/README.md:4]: I would just go with HPCS, with 1 or 2 lines of text explaining how to use key protect instead

File: examples/fscloud/main.tf [hunk 5]
```diff
@@ -84,12 +84,12 @@ module "cbr_account_level" {
...
   # This example open up:
-  #   1. Flows from icd mongodb, postgresql to kms on private endpoint
+  #   1. Flows from icd mongodb, postgresql to KMS and HPCS on private endpoint
   #   2. Flow from schematics on public kms endpoint
   #   3. Add a block of ips to schematics public endpoint
   #   4. Flow from vpc(s) specified in input zone_vpc_crn_list to postgresql private endpoint
-  custom_rule_contexts_by_service = {
-    "kms" = [{
+  custom_rule_contexts_by_service = merge({
+    for key in var.kms : key => [{
```
Comment (vburckhardt) [inline examples/fscloud/main.tf:92]: Is adding those rules in the example still relevant given that the root module add rules for all icd service now?

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Hi Akash, couple of comments from an initial quick pass.

File: examples/fscloud/variables.tf (hunk 1 above)
Comment (Ak-sky) [reply examples/fscloud/variables.tf:75]: Yes, my bad, resolved now.

File: modules/fscloud/variables.tf (hunk 3 above)
Comment (Ak-sky) [reply modules/fscloud/variables.tf:224]: ok resolved.

File: modules/fscloud/README.md (hunk 4 above)
Comment (Ak-sky) [reply modules/fscloud/README.md:4]: Done.

Comment (Ak-sky) [reply modules/fscloud/README.md:4]: Done.

File: examples/fscloud/main.tf (hunk 5 above)
Comment (Ak-sky) [reply examples/fscloud/main.tf:92]: Make sense, removed the rule for ICD.

**Comment by vburckhardt** [comment]
This PR will require a major release as there is a breaking change (change from key protect to HPCS).

The upgrade test should be failing in those scenarios but this will be requiring a bit of extra logic specific to CBR. The change here is only updating a list in an array (list of CBR context associated with a CBR rule) - so not being flagged with the generic update test logic.

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Hi Akash, there appears to be an issue.

On the fscloud example, try to run terraform apply on the main branch, then terraform apply -var="kms=[\"kms\"] against this branch. In theory the second apply should just create an empty hpcs rule with deny all and not touch the key protect rule. In practice, this is very different.

The other aspect I'd suggest to change would be on the value of the kms variable. I think key-protect should be used in the variable rather than kms to be very specific.

Terraform will perform the following actions:

  # module.cbr_account_level.module.cbr_rule["hs-crypto"].ibm_cbr_rule.cbr_rule will be created
  + resource "ibm_cbr_rule" "cbr_rule" {
      + created_at          = (known after apply)
      + created_by_id       = (known after apply)
      + crn                 = (known after apply)
      + description         = "fs-hs-crypto-rule"
      + enforcement_mode    = "report"
      + href                = (known after apply)
      + id                  = (known after apply)
      + last_modified_at    = (known after apply)
      + last_modified_by_id = (known after apply)
      + version             = (known after apply)

      + contexts {
          + attributes {
              + name  = "endpointType"
              + value = "public"
            }
          + attributes {
              + name  = "networkZoneId"
              + value = "2369e03926f9a6430962ec007a11dde4"
            }
        }
      + contexts {
          + attributes {
              + name  = "endpointType"
              + value = "private"
            }
          + attributes {
              + name  = "networkZoneId"
              + value = "9f1fa50c459d48b745b95a50b31081e2,7d5f79dc62880f3752426f11a01176bb,7d5f79dc62880f3752426f11a0116a5e,9f1fa50c459d48b745b95a50b3109da8,be6880d553a0b4014a7f6bf66e10a400,7d5f79dc62880f3752426f11a011650c,be6880d553a0b4014a7f6bf66e10b722,edd2f67507fc62c2e7647340bf11f804,2369e03926f9a6430962ec007a11f0de,9f1fa50c459d48b745b95a50b3108157,7d5f79dc62880f3752426f11a01158ef"
            }
        }

      + operations {
          + api_types {
              + api_type_id = "crn:v1:bluemix:public:context-based-restrictions::::api-type:"
            }
        }

      + resources {
          + attributes {
              + name     = "accountId"
              + operator = "stringEquals"
              + value    = "7aa6f7b185f2e3170fac9919aa1769ee"
            }
          + attributes {
              + name     = "serviceName"
              + operator = "stringEquals"
              + value    = "hs-crypto"
            }
        }
    }

  # module.cbr_account_level.module.cbr_rule["kms"].ibm_cbr_rule.cbr_rule will be updated in-place
  ~ resource "ibm_cbr_rule" "cbr_rule" {
        id                  = "650fae846e5cfd0850b0dd9aa711fcc7"
        # (9 unchanged attributes hidden)

      ~ contexts {
          ~ attributes {
                name  = "networkZoneId"
              ~ value = "9f1fa50c459d48b745b95a50b31081e2,7d5f79dc62880f3752426f11a01176bb,7d5f79dc62880f3752426f11a0116a5e,9f1fa50c459d48b745b95a50b3109da8,be6880d553a0b4014a7f6bf66e10a400,7d5f79dc62880f3752426f11a011650c,be6880d553a0b4014a7f6bf66e10b722,edd2f67507fc62c2e7647340bf11f804,2369e03926f9a6430962ec007a11f0de,9f1fa50c459d48b745b95a50b3108157,7d5f79dc62880f3752426f11a01158ef" -> "edd2f67507fc62c2e7647340bf11f1a5"
            }

            # (1 unchanged block hidden)
        }
      ~ contexts {
          ~ attributes {
                name  = "endpointType"
              ~ value = "private" -> "public"
            }
          ~ attributes {
                name  = "networkZoneId"
              ~ value = "edd2f67507fc62c2e7647340bf11f804,9f1fa50c459d48b745b95a50b3108157" -> "2369e03926f9a6430962ec007a11df44"
            }
        }
      - contexts {
          - attributes {
              - name  = "endpointType" -> null
              - value = "public" -> null
            }
          - attributes {
              - name  = "networkZoneId" -> null
              - value = "650fae846e5cfd0850b0dd9aa711cfac" -> null
            }
        }
      - contexts {
          - attributes {
              - name  = "endpointType" -> null
              - value = "public" -> null
            }
          - attributes {
              - name  = "networkZoneId" -> null
              - value = "2369e03926f9a6430962ec007a11df44" -> null
            }
        }

        # (3 unchanged blocks hidden)
    }

File: examples/fscloud/README.md [hunk 6]
```diff
@@ -3,17 +3,16 @@
...
 
 This examples is designed to show case some of the key customization options for the module. In addition to the pre-wired CBR rules documented at [fscloud profile](../../profiles/fscloud/), this examples show how to customize the module to:
-1. Open up network traffic flow from ICD mongodb, ICD Postgresql to the Key Protect private endpoints.
-2. Open up network traffic flow from Schematics to Key Protect public endpoints.
-3. Open up network traffic flow from a block of IPs to the Schematics public endpoint.
-4. Open up network traffic flow from the VPC created in this example to ICD postgresql private endpoints.
+1. Open up network traffic flow from Schematics to HPCS public endpoints.
+2. Open up network traffic flow from a block of IPs to the Schematics public endpoint.
+3. Open up network traffic flow from the VPC created in this example to ICD postgresql private endpoints.
+4. Open up network traffic flow from RabbitMQ to the Key Protect private endpoints.
```
Comment (vburckhardt) [inline examples/fscloud/README.md:9]: I'd suggest leaving all ICD -> KMS flow open in fscloud module here, rather than just opening up RabbitMQ specifically in this example.

**Review by vburckhardt — CHANGES_REQUESTED** [review]

**Comment by Ak-sky** [comment]
Hi Vincent, thank you for the review.
I have below few points which I want to discuss- 

- The second apply is touching the key protect rule because there are changes in the FScloud example (current working branch) and due to this updates are being applied to the KMS rule.

- From main branch we are setting the `custom_rule_contexts_by_service` as below from `main` branch
   

    ```
    custom_rule_contexts_by_service = {
        "kms" = [{
          endpointType      = "private"
          service_ref_names = ["databases-for-mongodb", "databases-for-postgresql"]
          },
          {
            endpointType      = "public"
            service_ref_names = ["schematics"]
          },
          {
            endpointType = "public"
          zone_ids = [module.cbr_zone_operator_ips.zone_id] }
        ],
        "schematics" = [{
          endpointType = "public"
          zone_ids     = [module.cbr_zone_operator_ips.zone_id]
        }],
        "databases-for-postgresql" = [{
          endpointType = "private"
          ## Give access to the zone containing the VPC passed in zone_vpc_crn_list input
          add_managed_vpc_zone = true
        }]
      }
    ```
    and from working branch there is little change in `custom_rule_contexts_by_service` where var.kms = kms, it is detecting a change.
    
        
          custom_rule_contexts_by_service = merge({
            for key in var.kms : key => [{
              endpointType      = "public"
              service_ref_names = ["schematics"]
              },
              {
                endpointType = "public"
              zone_ids = [module.cbr_zone_operator_ips.zone_id] }
            ] }, {
            "schematics" = [{
              endpointType = "public"
              zone_ids     = [module.cbr_zone_operator_ips.zone_id]
            }],
            "databases-for-postgresql" = [{
              endpointType = "private"
              ## Give access to the zone containing the VPC passed in zone_vpc_crn_list input
              add_managed_vpc_zone = true
            }]
            }, {
            "kms" = [{
              endpointType      = "private"
              service_ref_names = ["messages-for-rabbitmq"]
              },
              {
                endpointType = "public"
              zone_ids = [module.cbr_zone_operator_ips.zone_id] }
          ] })
        
- To test this change in the FScloud example, I did apply on the `main` branch but copied the `custom_rule_contexts_by_service` as below from working branch and did the second TF apply, there I could see it was only creating one rule for HPCS with deny all only.
```
       custom_rule_contexts_by_service = {
    "kms" = [{
      endpointType      = "public"
      service_ref_names = ["schematics"]
      },
      {
        endpointType = "public"
      zone_ids = [module.cbr_zone_operator_ips.zone_id] }
    ],
    "schematics" = [{
      endpointType = "public"
      zone_ids     = [module.cbr_zone_operator_ips.zone_id]
    }],
    "databases-for-postgresql" = [{
      endpointType = "private"
      ## Give access to the zone containing the VPC passed in zone_vpc_crn_list input
      add_managed_vpc_zone = true
    }],
    "kms" = [{
      endpointType      = "private"
      service_ref_names = ["messages-for-rabbitmq"]
      },
      {
        endpointType = "public"
      zone_ids = [module.cbr_zone_operator_ips.zone_id] }
  ] }
```
```
  Terraform will perform the following actions:

  # module.cbr_account_level.module.cbr_rule["hs-crypto"].ibm_cbr_rule.cbr_rule will be created
  + resource "ibm_cbr_rule" "cbr_rule" {
      + created_at          = (known after apply)
      + created_by_id       = (known after apply)
      + crn                 = (known after apply)
      + description         = "fs-03-hs-crypto-rule"
      + enforcement_mode    = "report"
      + href                = (known after apply)
      + id                  = (known after apply)
      + last_modified_at    = (known after apply)
      + last_modified_by_id = (known after apply)
      + version             = (known after apply)

      + contexts {
          + attributes {
              + name  = "endpointType"
              + value = "public"
            }
          + attributes {
              + name  = "networkZoneId"
              + value = "afe33d0061b6cea2c2e6f54311210501"
            }
        }

      + operations {
          + api_types {
              + api_type_id = "crn:v1:bluemix:public:context-based-restrictions::::api-type:"
            }
        }

      + resources {
          + attributes {
              + name     = "accountId"
              + operator = "stringEquals"
              + value    = "abac0df06b644a9cabc6e44f55b3880e"
            }
          + attributes {
              + name     = "serviceName"
              + operator = "stringEquals"
              + value    = "hs-crypto"
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  ~ map_target_service_rule_ids     = {
      + hs-crypto                   = {
          + rule_id = (known after apply)
        }
        # (28 unchanged attributes hidden)
    }
```
  - Verified from the UI as well-
     
     image

File: examples/fscloud/README.md (hunk 6 above)
Comment (Ak-sky) [reply examples/fscloud/README.md]: Done.

**Review by vburckhardt — CHANGES_REQUESTED** [review]
Hi Akash, the reverse is needed on the renaming. This variable is meant to enable to select the kms (key management system) to either key protect or hpcs.

variable "kms" {
  type        = list(string)
  description = "List specific Key Management Service (KMS) type, valid values are 'key-protect' for Key Protect and 'hs-crypto' for HPCS"
  default     = ["hs-crypto"]
  validation {
    condition = alltrue([
      for key_protect_val in var.key_protect : can(regex("^(key-protect|hs-crypto)$", key_protect_val))
    ])
    error_message = "Valid values for key_protect are 'key-protect' for Key Protect and 'hs-crypto' for HPCS"
  }
}

**Comment by Ak-sky** [comment]
> Hi Akash, the reverse is needed on the renaming. This variable is meant to enable to select the kms (key management system) to either key protect or hpcs.
> 
> ```
> variable "kms" {
>   type        = list(string)
>   description = "List specific Key Management Service (KMS) type, valid values are 'key-protect' for Key Protect and 'hs-crypto' for HPCS"
>   default     = ["hs-crypto"]
>   validation {
>     condition = alltrue([
>       for key_protect_val in var.key_protect : can(regex("^(key-protect|hs-crypto)$", key_protect_val))
>     ])
>     error_message = "Valid values for key_protect are 'key-protect' for Key Protect and 'hs-crypto' for HPCS"
>   }
> }
> ```


Hi Vincent, 

Valid target service name that is supported by CBR for `Key Protect` is `kms` hence `key-protect` can't be used as kms variable value.

Below is my suggestion-

```
variable "key_management_system" {
  type        = list(string)
  description = "List specific Key Management Service (KMS) type, valid values are 'kms' for Key Protect and 'hs-crypto' for HPCS"
  default     = ["hs-crypto"]
  validation {
    condition = alltrue([
      for kms_val in var.key_management_system : can(regex("^(kms|hs-crypto)$", kms_val))
    ])
    error_message = "Valid values for key_management_system are 'kms' for Key Protect and 'hs-crypto' for HPCS"
  }
}
```

**Comment by vburckhardt** [comment]
> Valid target service name that is supported by CBR for Key Protect is kms hence key-protect can't be used as kms variable value.

I was seeing the module being responsible for mapping the key-protect input to kms value when creating the CBR via the terraform provider. There is a naming issue at the CBR stack (kms instead of key-protect) that I think the module should attempt to fix here.

File: modules/fscloud/README.md [hunk 7]
```diff
@@ -1,16 +1,18 @@
...
+- ROKS -> HPCS
 - Activity Tracker route -> COS
 - VPCs where clusters are deployed -> COS
 - IS (VPC Infrastructure Services) -> COS
 - VPCs -> container registry
-- All ICD -> KMS
+- All ICD -> HPCS
 - IKS -> IS (VPC Infrastructure Services)
 
+This module now supports Hyper Protect Crypto Services. By default it is set to use Hyper Protect Crypto Services but this can be modified to use Key Protect or both Key Protect and Hyper Protect Crypto Services using the terraform variable `key_protect`.
```
Comment (vburckhardt) [inline modules/fscloud/README.md:14]: the name of the variable is 'kms' though? key-protect is the value.

**Review by vburckhardt — CHANGES_REQUESTED** [review]

File: modules/fscloud/README.md (hunk 7 above)
Comment (Ak-sky) [reply modules/fscloud/README.md:14]: Done.

File: examples/fscloud/main.tf [hunk 8]
```diff
@@ -86,23 +93,19 @@ module "cbr_account_level" {
...
-      endpointType      = "private"
-      service_ref_names = ["databases-for-mongodb", "databases-for-postgresql"]
-      },
-      {
-        endpointType      = "public"
-        service_ref_names = ["schematics"]
+  #   1. Flow from schematics on public HPCS endpoint
+  #   2. Add a block of ips to schematics public endpoint
+  #   3. Flow from vpc(s) specified in input zone_vpc_crn_list to postgresql private endpoint
+  #   4. Flows from RabbitMQ to KMS on private endpoint
```
Comment (vburckhardt) [inline examples/fscloud/main.tf:99]: I believe you meant to comment about Postgresql and not rabbitMQ? There is no trace of RabbitMQ in the code.

**Review by vburckhardt — CHANGES_REQUESTED** [review]

File: examples/fscloud/main.tf (hunk 8 above)
Comment (Ak-sky) [reply examples/fscloud/main.tf]: I agree Vincent, I have made changes to the comments accordingly.

**Review by vburckhardt — APPROVED** [review]
